File: Java/Merge_K_SortedLists.java
/**Problem Desciption
 *  You are given an array of k linked-lists lists, each linked-list is sorted in ascending order.
    Merge all the linked-lists into one sorted linked-list and return it.

    Problem link - https://leetcode.com/problems/merge-k-sorted-lists/
 */


/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Merge_K_SortedLists {
    public ListNode mergeKLists(ListNode[] lists) {
        int k=lists.length;
        ListNode sorted=new ListNode(0);
        if(lists.length==0)
            return sorted.next;
        PriorityQueue<ListNode> pq=new PriorityQueue<>(k,(a,b)->Integer.compare(a.val,b.val));
        ListNode ptr=sorted;
        for(int i=0;i<k;i++)
            if(lists[i]!=null)
                pq.add(lists[i]);
        while(!pq.isEmpty())
        {
            ListNode temp=pq.poll();
            ptr.next=temp;
            ptr=ptr.next;
            if(temp.next!=null)
            {
                temp=temp.next;
                pq.add(temp);
            }
            
        }
        return sorted.next;
    }
    
} 

File: Python/Merge_k_Sorted_Lists.py
# Problem Link: https://leetcode.com/problems/merge-k-sorted-lists/

from heapq import merge


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:

    def mergeKLists(self, lists: List[ListNode]) -> ListNode:
        lists = list(filter(lambda x: x is not None, lists))

        array_list = []
        array = []
        for l in lists:
            node = l

            while node:
                array.append(node.val)
                node = node.next

            array_list.append(array)
            array = []

        if (len(array_list) > 0):
            merge_list = array_list[0]

            for i in range(len(array_list) - 1):
                merge_list = list(merge(merge_list, array_list[i + 1]))

            linked_list = output = ListNode(int(merge_list[0]))

            for j in range(len(merge_list) - 1):
                output.next = ListNode(int(merge_list[j + 1]))
                output = output.next

            return linked_list
        else:
            linked_list = output = ListNode(None)
            return output.next